### ISSUE TITLE
StreamField should consistently use bulk_to_python() to improve efficiency

### ISSUE THREAD

**Issue by ababic** [wagtail/wagtail#5926]
### Is your proposal related to a problem?

It's very common in projects that I work on to define blocks that have a heading of some kind, followed by a curated list of pages, snippets or other objects (maybe up to 10 or so). Currently, when rendering a page with such a block in place, as many queries as there are items are made to fetch the necessary data, which clearly isn't as efficient as it could be.

### Describe the solution you'd like

Nested instances of ListBlock should utilise `bulk_to_python()` to fetch multiple objects at once.

An added bonus would be adding some simple way of optimising further, by allowing some control over how 'prefetch_related' and 'select_related' are used in `bulk_to_python()`. Perhaps they could be supplied as arguments, or each block could look for `streamvalue_select_related` and `streamvalue_prefetch_related` attributes on the target model, or a `for_streamvalue()` method on the queryset itself?

### Additional context

It looks as though `StreamValue`[ should already support this](https://github.com/wagtail/wagtail/blob/e1d3390a1f62ae4b30c0ef38e7cfa5070d8565dc/wagtail/core/blocks/stream_block.py#L408), but it seems not to be working past the first level for me when testing locally. The built-in `SnippetChooserBlock`, `PageChooserBlock`, `ImageChooserBlock` and `DocumentChooserBlock` classes already implement a `bulk_to_python()` method to support such optimisation. 

Below is a generic block example that can be used to test this:

```
from wagtail.images.blocks import ImageChooserBlock
from wagtail.core.blocks import CharBlock, ListBlock, StructBlock

class GalleryBlock(StructBlock):
    heading = CharBlock(required=False)
    images = ListBlock(ImageChooserBlock())
```

**Comment by chosak** [comment]
I've opened #5930 that partially addresses this by allowing each ListBlock to retrieve its items in a single query. This doesn't address the more general desire to do bulk loading across ListBlocks; for example, in the example you provide above @ababic it'd be nice to be able to load _all_ images across _all_ GalleryBlocks at once.

To do that I think we'd need a way to first recursively traverse the tree to find all the child blocks of a certain type and retrieve them together (which feels like it'd overlap with the logic needed to do #4298 correctly). This gets tricky as the blocks get more complex; as an example, see [this](https://github.com/cfpb/cfgov-refresh/blob/1b478eb9fe2b61ea29a344e898028da4f92130eb/cfgov/mega_menu/blocks.py#L29) code that tries to do something similar. First you have to traverse the block to find all page references, then retrieve all pages at once, then go back and replace the references with the page objects. I'm not sure what the best generic approach to do this would be.

**Comment by ababic** [comment]
Thanks @chosak. I'll take progress over perfection any day, and the changes in #5930 (small as they are) will already make a huge improvement.

If we were to tackle generic prefetching of all values, I wonder whether it's possible to collect all model instances into a dictionary on the root element, then have chooser blocks automatically draw their value from that somehow?  To do that, I guess all blocks would have to have knowledge of all of their ancestors and continue to pass that data down the chain.

**Comment by gasman** [comment]
The original plan for `bulk_to_python` was that all container-type blocks would implement it recursively - so, for example, `ListBlock.bulk_to_python` (which applies `to_python` to a list of lists) would concatenate its input into a single list (keeping track of the original list lengths), make a single call to `child_block.bulk_to_python`, and split the results into sub-lists again. Meanwhile, `StructBlock.bulk_to_python` would organise its input (a list of dicts) into one list per child field, run each of those lists through the respective child block's `bulk_to_python`, and then shuffle everything back into dicts.

This would allow `bulk_to_python` to work over a StreamField structure of any depth and complexity, making only one call to `bulk_to_python` for each distinct block definition in the StreamField, while preserving the important design principle that no block type needs to know the "big picture" of how other containers work, and the root element doesn't have to be a special case.

I thought we had a naive implementation of `bulk_to_python` on the base `Block` class, that simply iterated over the list calling `self.to_python` on it, but I guess we didn't get that far - I think that would be the next step towards the full implementation, as it saves us from having to stick `if hasattr(self.child_block, 'bulk_to_python')` branches all over the place.

**Comment by ababic** [comment]
@gasman Thanks for outlining. This approach doesn't sound a million miles away from what @chosak was suggesting. Am I correcting in thinking this could be done in a backwards-compatible way, and wouldn't be a horrendous amount of work?

**Comment by gasman** [comment]
@ababic Yep - pretty sure it should just fall into place once all the blocks consistently implement the new method, no major architectural changes required.

**Comment by chosak** [comment]
@ababic @gasman I've merged #5930 in 3797132b4, which partially addresses the title of this issue although not the more general description and goal of more universal bulk_to_python support. Would it make sense to leave this open, or close and replace with another more generically-written issue?

**Comment by ababic** [comment]
Thanks @chosak. I have renamed this issue to make it a little more generic, with the idea we continue to use it to track further progress.

**Comment by gasman** [comment]
Fixed in #5976
